0012-five-bells-ledger-api/0012-five-bells-ledger-api.md

```diff
@@ -0,0 +1,1290 @@
+
+`^[-+]?[0-9]*[.]?[0-9]+([eE][-+]?[0-9]+)?$`
+
+Client applications can decode numeric strings to whatever representation provides sufficient precision for their specific needs. The ledger MUST report information about the precision it uses in the [Get Ledger Metadata][] response.
+
+
+### Assets and Currency
+
+The Five Bells Ledger API provides access to a ledger containing a single currency or asset. The [Get Ledger Metadata][] method reports the type of asset or currency used.
+
+If a provider supports multiple currencies or assets, the provider can serve Five Bells Ledger APIs separately for each such asset by serving it from a different prefix. For example, the [Prepare Transfer][] method could be available at all the following locations:
+
+- `POST https://ledger.example.com/TZS/transfers`
+- `POST https://ledger.example.com/KES/transfers`
+- `POST https://ledger.example.com/stocks/MSFT/transfers`
+
+Any valid URL is a suitable prefix. However, only the same asset served by the same underlying ledger can be directly transferred.
+
+These instances could be served by the same underlying software and share infrastructure, but it would not be possible to transfer directly from one to the other, since the currency denominations are different. An **ILP Connector** could facilitate cross-currency payments.
+
+#### Adding Money to the Ledger
+
+There are several ways to "add" money to the ledger. A ledger MAY use any or all of the following:
+
+- When an account is created or updated by an administrator with the [Create or Update Account][] method, the administrator MAY set the balance of the account to any legal value.
+- A ledger may be initialized in a state where its database contains accounts with positive balances.
+- An administrator MAY create an "issuer" account whose balance can be negative. To give a positive balance to an account, execute a transfer from the issuer account. If you use this technique exclusively, the issuer account's total balance is equal to the negative of all balances in the ledger summed together.
+
+
+### Scope
+
+The Five Bells Ledger API defines one way to implement sufficient functionality to operate a ledger. It is not intended to be an optimal or ideal ledger, especially since such a thing is subjective or at least hard to quantify. (Indeed, Interledger itself was conceived from the conclusion that no one ledger can serve the needs of every person in existence.) In general, this specification reflects to the existing [Five Bells Ledger reference implementation](https://github.com/interledgerjs/five-bells-ledger) so that others may implement compatible software.
+
+
+### Authorization and Authentication
+
+The ledger MUST authenticate API clients. HTTP(S) methods MUST use one of the following systems for authentication:
+
+- [HTTP Basic Auth](https://tools.ietf.org/html/rfc2617#section-2)
+- [TLS Client Certificates](https://tools.ietf.org/html/rfc5246#section-7.4.6)
```

Comment:
It is for the ledger computational expensive to verify that credentials passed in via Basic Auth are correct (it should take ~1 sec. for security reasons). In contrast, token-based authentication (e.g. JSON Web Tokens in fb-ledger) are much cheaper to verify for the server.

By switching from Basic Auth to token-based authentication we make it more difficult for unauthorized requests to put a substantial load on the server. Only the endpoint that hands out the auth token should be callable with HTTP Basic Auth. Furthermore, this endpoint should have a strict rate-limiting policy to prevent flooding with fake credentials.